### ISSUE TITLE
Use new (yet-to-be-created) netcdf4-native

### ISSUE THREAD

**Issue by lesserwhirls** [Unidata/netcdf-java#112]
* Bundle netcdf-c native libs for mac, windows, linux in single jar.
* Extend netcdf4 module code to make one last attempt to find a suitable library on the classpath (using [JNA extractResourceFromPath](http://java-native-access.github.io/jna/5.4.0/javadoc/com/sun/jna/Native.html#extractFromResourcePath-java.lang.String-)).

**Comment by lesserwhirls** [comment]
@rschmunk - based on your experience over on Unidata/netcdf-java#1161, what would be the minimum set of features you'd need to have the HDF5 and netCDF-C libraries built with to be shipped with a `netcdf4-native` jar (much like the [libaec-native](https://github.com/Unidata/netcdf-java/issues/753) jar)?

**Comment by rschmunk** [comment]
@lesserwhirls , To be honest, I have no idea. I only get into this stuff when I hear from people that Panoply (and thus netCDF-Java) won't open their datasets.

The libaec case was the bigger one, as was demonstrated by the number of people chiming in on #753 that they couldn't open their ECMWF data product.

With #1161, It's only been just a few people asking about CDF5/Pnetcdf datasets over the past couple years, but having just heard from another a few days ago I figured it was time to figure the thing out. I was pleasantly surprised how easy that was to resolve once I tracked down the bug in #1473. OTOH, am I going to have to provide instructions on how to set up an nj22config.xml file to every such person in future?

Currently, my only noteworthy concern about being able to open/read dataset formats has to do with Zarr and filters, per #1447. It's not a pressing issue, but I'm not going to advertise Zarr capability much until the capability improves.

Aside from that, the main thing I'm presently trying to get my head (and code) around is accessing structures and more specifically structure member values so that they can be plotted. Lately I have been seeing more sample remote sensing data products where complex numbers were stored as structures with real and imaginary members.

Edit: And once I get a better grip on extracting structure member values cleanly, I'll be asking again about obtaining coordinate systems for the structures, as that's not available in 5.9. But maybe is in 6.0? See #1378.

**Comment by rschmunk** [comment]
Something to consider is whether attempting to load an included netCDF-C native library could be made optional.

Given the problem described over in #1475 , it would be useful to be able to disable such loading if it seems like might be causing opening a dataset to fail.

**Comment by lesserwhirls** [comment]
For sure. Reading through the C library would be an opt-in thing through the existing mechanisms. The main reason to bundle the C library in a jar would be ease of use for writing netCDF-4 files.

**Comment by rschmunk** [comment]
Hmmm, since my code doesn't need to write netCDF-4 files, I wouldn't necessarily even need the netCDF-C library to be present at all. It's a trade-off btw how much does it complicate my build process, how much does it increase the size of my app distro, and just how many users need its ability to read some things that netCDF-Java currently cannot.

My next Panoply release makes it optional as to whether to read the nj22Config.xml file if it is present, so the user can just flip a switch in the prefs UI if they think that using the C library is causing trouble, a la #1475 . I probably need a help page explaining what that bit is all about.

**Comment by Yaqiang** [comment]
netCDF-4 file writing is important for most advanced users,  it's definitely useful to bundle the C library in a jar with netCDF java package.

**Comment by rschmunk** [comment]
A user just wrote in about a warning message he's seeing for the first time, and at the moment it looks like it's triggered when the `NetcdfClibrary` class calls `Native.load()`.

```
WARNING: A restricted method in java.lang.System has been called
WARNING: java.lang.System::load has been called by com.sun.jna.Native in an unnamed module (file:/some/path/PanoplyJ/jars/netcdfAll-5.9.1-SNAPSHOT.jar)
WARNING: Use --enable-native-access=ALL-UNNAMED to avoid a warning for callers in this module
WARNING: Restricted methods will be blocked in a future release unless native access is enabled
```

This is due to changes made in Java 24. Details [here](https://inside.java/2024/12/09/quality-heads-up/) .

What's going on in Java is plain enough, but I am confused at the moment what netCDF-Java code is trying to use the `com.sun.jna.Native` class. The user does not have an nj22Config.xml file, so Panoply isn't deliberately trying to load the netCDF-C library. Is NJ trying to do so on its own? The user is apparently on a Linux supercomputer, so there may well be pertinent environment variables set that I never deal with when doing stuff on a Mac.

**Comment by lesserwhirls** [comment]
I've been working on upgrading our gradle build to gradle 9 (and migrating to the Kotlin DSL for configuration of the build), and only yesterday encountered this for the first time while running tests using JDK 25. Are you bundling the `libaec-native` subproject project into `netcdfAll.jar` to handle CCSDS compression? If so, that will be calling JNA, and could likely be the [trigger for this warning](https://java-native-access.github.io/jna/5.18.1/javadoc/index.html#special-considerations-for-jdk24--heading).

**Comment by rschmunk** [comment]
Yes, I'm including` libaec-native` in the jar when building netCDF-Java. However, if I set Panoply to **_not_** load the netCDF-C library at launch and then drag and drop one of the problem GRIB DRS 42 ECMWF datasets into the app, I don't get the above warning. Instead I get…

```
WARNING: A restricted method in java.lang.System has been called
WARNING: java.lang.System::loadLibrary has been called by gov.nasa.giss.macos.MacNative in an unnamed module (file:/some/path/PanoplyJ/jars/netcdfAll-5.9.2-SNAPSHOT.jar)
WARNING: Use --enable-native-access=ALL-UNNAMED to avoid a warning for callers in this module
WARNING: Restricted methods will be blocked in a future release unless native access is enabled
```

Note that it says `System::loadLibrary` was called, not `System::load` as in my prior comment.

Oh, and when working with that DRS42 file, I subsequently also get…

```
WARNING: A terminally deprecated method in sun.misc.Unsafe has been called
WARNING: sun.misc.Unsafe::arrayBaseOffset has been called by com.google.protobuf.UnsafeUtil$MemoryAccessor (file:/some/path/PanoplyJ/jars/netcdfAll-5.9.2-SNAPSHOT.jar)
WARNING: Please consider reporting this to the maintainers of class com.google.protobuf.UnsafeUtil$MemoryAccessor
WARNING: sun.misc.Unsafe::arrayBaseOffset will be removed in a future release
```

**Comment by rschmunk** [comment]
Summing up my experience with J24+ so far, there are three things happening I've seen or had reported with Panoply+NJ.

1) On my Mac, I configure Panoply to parse an `nj22Config.xml` file at launch and then load netCDF-C. Immediate logging is the `Save::load` warning.

2) On My Mac, I don't do that configuration, but then I open a GRIB DRS 42 file. Logging then reports the `Save::loadLibrary` warning.

3) A Panoply user tells me he's getting the `Save::load` warning even though he doesn't have a `nj22Config.xml` file and he's not working with a GRIB file. I'm still puzzling out WTF this might be. Wondering if it's some obscure hole in my code or if it's because he's on a different OS and his environment variables are triggering "something" in NJ.

I'm hoping that once I can provide the user in (3) with a beta or new Panoply release after the gov't shutdown concludes that he'll no longer see the (3) warning. But at the moment, IDK.

**Comment by lesserwhirls** [comment]
Thank you for outlining your experience with J24+ @rschmunk! So I can see two things happening here.

First, there is something triggering the `java.lang.System::load` warning when netCDF-Java should not, in theory, be trying to load libaec or netCDF-C. That warning comes from JNA, which is what we use to interface those two C libraries. I think JNA is used by one of our 3rd party dependencies, so it might be that library calling it and not us directly, but I'm not sure. I'll need to dig into this one further to see exactly what is trigger it. A JNA or JNI call is also likely responsible for the `java.lang.System::loadLibrary` warning, but again, I'll need to dig into things to see exactly what is trigger that.

Second, it seems the protobuf library is doing some calls that trigger a `sun.misc.Unsafe` warning. We use protobuf when reading/writing GRIB index files, as well as a few other places in the library, so that's probably why you see that message when trying to load a GRIB file (any GRIB file should trigger that one). The protobuf library devs are aware of it, and are [looking into ways of addressing it](https://github.com/protocolbuffers/protobuf/issues/20760#issuecomment-3304650765).
